joern-cli/frontends/rubysrc2cpg/src/main/scala/io/joern/rubysrc2cpg/astcreation/AstCreator.scala

```diff
@@ -37,18 +48,108 @@ class AstCreator(filename: String, global: Global)
+  private val methodNameAsIdentiferQ = mutable.Queue[Ast]()
+
+  private val methodAliases = mutable.HashMap[String, String]()
+
+  class ScopeIdentifiers {
+    val varToIdentiferMap             = mutable.HashMap[String, NewIdentifier]()
+    var parentScope: ScopeIdentifiers = null
+  }
+
+  private val scopeStack = mutable.Stack[ScopeIdentifiers]()
+
+  private def pushScope(isParentAccesible: Boolean = false): ScopeIdentifiers = {
+    val scope = new ScopeIdentifiers()
+
+    if (isParentAccesible) {
+      scope.parentScope = scopeStack.top
+    }
+    scopeStack.push(scope)
+    scope
+  }
+
+  private def popScope(): Unit = {
+    scopeStack.pop()
+  }
+
+  private def setIdentiferInScope(node: NewIdentifier): Unit = {
+    scopeStack.top.varToIdentiferMap.getOrElseUpdate(node.name, node)
+  }
+
+  private def lookupIdentiferInScope(name: String): Boolean = {
+    scopeStack.top.varToIdentiferMap.contains(name)
+  }
+
+  private def createIdentiferWithScope(
+    node: TerminalNode,
+    name: String,
+    code: String,
+    typeFullName: String,
+    dynamicTypeHints: Seq[String]
+  ): NewIdentifier = {
+    val newNode = identifierNode(node, name, code, typeFullName, dynamicTypeHints)
+    setIdentiferInScope(newNode)
+    newNode
+  }
+
+  private def getActualMethodName(name: String): String = {
+    methodAliases.getOrElse(name, name)
+  }
   override def createAst(): BatchedUpdate.DiffGraphBuilder = {
     val charStream  = CharStreams.fromFileName(filename)
     val lexer       = new RubyLexer(charStream)
     val tokenStream = new CommonTokenStream(lexer)
     val parser      = new RubyParser(tokenStream)
     val programCtx  = parser.program()
 
-    val statementCtx   = programCtx.compoundStatement().statements()
-    val statementAst   = astForStatementsContext(statementCtx)
+    val statementCtx = programCtx.compoundStatement().statements()
+    pushScope()
+    val statementAsts = if (statementCtx != null) {
+      astsForStatementsContext(statementCtx)
+    } else {
+      List[Ast](Ast())
+    }
+    popScope()
+
+    val name = ":program"
+    val programMethod =
+      NewMethod()
+        .order(1)
+        .name(name)
+        .code(name)
+        .fullName(filename)
+        .filename(filename)
+        .lineNumber(0)
+        .lineNumberEnd(-1)
+        .columnNumber(0)
+        .columnNumberEnd(-1)
+        .astParentType(NodeTypes.TYPE_DECL)
+        .astParentFullName(filename)
+
+    val thisParam = NewMethodParameterIn()
+      .name("this")
+      .code("this")
+      .lineNumber(0)
+      .columnNumber(0)
+    val thisParamAst = Ast(thisParam)
+
+    val methodRetNode = NewMethodReturn()
+      .lineNumber(None)
+      .columnNumber(None)
+      .typeFullName(Defines.Any)
+
+    val blockNode = NewBlock().typeFullName(Defines.Any)
+    val programAst =
+      methodAst(programMethod, Seq[Ast](thisParamAst), blockAst(blockNode, statementAsts), methodRetNode)
```

Comment:
You can probably leave out the `[Ast]` here.

Reply:
Done for this case and all other cases